board/safety/safety_mazda.h

```diff
@@ -21,56 +22,125 @@
 #define MAZDA_MAX_RATE_DOWN 25
 #define MAZDA_DRIVER_TORQUE_ALLOWANCE 15
 #define MAZDA_DRIVER_TORQUE_FACTOR 1
+#define MAZDA_MAX_TORQUE_ERROR 350
 
-int mazda_cruise_engaged_last = 0;
-int mazda_rt_torque_last = 0;
-int mazda_desired_torque_last = 0;
-uint32_t mazda_ts_last = 0;
-struct sample_t mazda_torque_driver;         // last few driver torques measured
+// lkas enable/disbale speeds 51.5kph and 45kph
+#define MAZDA_LKAS_ENABLE_SPEED  5150
+#define MAZDA_LKAS_DISABLE_SPEED 4500
+
+const CanMsg MAZDA_TX_MSGS[] = {{MAZDA_LKAS, 0, 8}};
+bool mazda_lkas_on = false;
+
+AddrCheckStruct mazda_rx_checks[] = {
+  {.msg = {{MAZDA_CRZ_CTRL,     0, 8}}, .expected_timestep = 20000U},
+  {.msg = {{MAZDA_WHEEL_SPEED,  0, 8}}, .expected_timestep = 10000U},
+  {.msg = {{MAZDA_STEER_TORQUE, 0, 8}}, .expected_timestep = 12000U},
+  {.msg = {{MAZDA_ENGINE_DATA,  0, 8}}, .expected_timestep = 10000U},
+  {.msg = {{MAZDA_PEDALS,       0, 8}}, .expected_timestep = 20000U},
+};
+const int MAZDA_RX_CHECKS_LEN = sizeof(mazda_rx_checks) / sizeof(mazda_rx_checks[0]);
 
 // track msgs coming from OP so that we know what CAM msgs to drop and what to forward
 static int mazda_rx_hook(CAN_FIFOMailBox_TypeDef *to_push) {
-  int bus = GET_BUS(to_push);
-  int addr = GET_ADDR(to_push);
+  bool valid;
 
-  if ((addr == MAZDA_STEER_TORQUE) && (bus == MAZDA_MAIN)) {
-    int torque_driver_new = GET_BYTE(to_push, 0) - 127;
-    // update array of samples
-    update_sample(&mazda_torque_driver, torque_driver_new);
-  }
+  valid = addr_safety_check(to_push, mazda_rx_checks, MAZDA_RX_CHECKS_LEN,
+                            NULL, NULL, NULL);
+  if (valid) {
+    int bus = GET_BUS(to_push);
+    int addr = GET_ADDR(to_push);
+
+    if (bus == MAZDA_MAIN) {
+
+      // sample speed
+      if (addr == MAZDA_WHEEL_SPEED) {
+        int speed = 0;
+        // sum 4 wheel speeds, speed signal offset is 10000 with a 0.01 scale
+        for (int i=0; i<8; i+=2) {
+          int next_byte = i + 1;  // hack to deal with misra 10.8
+          speed += (GET_BYTE(to_push, i) << 8) + GET_BYTE(to_push, next_byte) - 10000;
+        }
+        speed = speed / 4;
+        vehicle_moving = ABS(speed) > 10; // moving is the speed is > 0.1 kph
+
+        // Enable LKAS at 51.5kph going up, disable at 45kph going down
+        if (speed > MAZDA_LKAS_ENABLE_SPEED) {
+          mazda_lkas_on = true;
+        } else if (speed < MAZDA_LKAS_DISABLE_SPEED) {
+          mazda_lkas_on = false;
+        } else {
+          // Misra-able appeasment block!
+        }
```

Comment:
this can just be `mazda_lkas_on  = speed > MAZDA_LKAS_ENABLE_SPEED` since we only check this when initially engaging